napari/plugins/utils.py

```diff
@@ -0,0 +1,153 @@
+from . import PluginError, log_plugin_error
+from . import plugin_manager as napari_plugin_manager
+from ._hookexec import _hookexec
+from typing import Union, List, Set
+from types import ModuleType
+from pluggy.hooks import HookImpl, _HookCaller
+
+HookOrderType = Union[List[str], List[ModuleType], List[HookImpl]]
+
+
+def permute_hook_implementations(
+    hook_caller: _HookCaller, order: HookOrderType
+):
+    """Change the call order of hookimplementations for a pluggy HookCaller.
+
+    Pluggy does not allow a built-in way to change the call order after
+    instantiation.  hookimpls are called in last-in-first-out order.
+    This function accepts the desired call order (a list of plugin names, or
+    plugin modules) and reorders the hookcaller accordingly.
+
+    Parameters
+    ----------
+    hook_caller : pluggy.hooks._HookCaller
+        The hook caller to reorder
+    order : list
+        A list of str, hookimpls, or module_or_class, with the desired
+        CALL ORDER of the hook implementations.
+
+    Raises
+    ------
+    ValueError
+        If the 'order' list cannot be interpreted as a list of "plugin_name"
+        or "plugin" (module_or_class)
+    ValueError
+        if 'order' argument has multiple entries for the same hookimpl
+    """
+    if all(isinstance(o, HookImpl) for o in order):
+        attr = None
+    elif all(isinstance(o, str) for o in order):
+        attr = 'plugin_name'
+    elif any(isinstance(o, str) for o in order):
+        raise TypeError(
+            "order list must be either ALL strings, or ALL modules/classes"
+        )
+    else:
+        attr = 'plugin'
+
+    hookimpls = hook_caller.get_hookimpls()
+    if len(order) > len(hookimpls):
+        raise ValueError(
+            f"too many values ({len(order)} > {len(hookimpls)}) in order."
+        )
+    if attr:
+        hookattrs = [getattr(hookimpl, attr) for hookimpl in hookimpls]
+    else:
+        hookattrs = hookimpls
+
+    # find the current position of items specified in `order`
+    indices = []
+    seen: Set[HookOrderType] = set()
+    for i in order:
+        if i in seen:
```

Comment:
I find the extra `seen` check within this loop a little confusing. You could add an assert check at the beginning of this method below line 52 that raises a ValueError if `len(your_list) != len(set(your_list))`. This isn't really a optimization or anything though just my personal preference so feel free to ignore : )

Reply:
yeah that's better. I think when I was writing this there was some reason why I was trying not to iterate that list too many times (and a greedy check like that causes one more loop right?) ... but now I can't remember why that was! 😅